QemuPkg/FrontPageButtons/FrontPageButtons.c

```diff
@@ -0,0 +1,253 @@
+/** @file FrontpageButtons.c
+
+ This module installs the MsButtonServicesProtocol and reports the requested state of Vol+ and
+ Vol- as indicated by the smbios table type3 record version string.
+
+ If the Type3.version string is:
+     "Vol+"  -- Return Vol+ pressed.
+     "Vol-"  -- Return Vol- pressed.
+     any other string, or no string, return neither pressed
+
+  Use only for Qemu platforms which can change smbios every time it boots.
+
+  Copyright (C) Microsoft Corporation. All rights reserved.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include <Uefi.h>
+
+#include <Protocol/ButtonServices.h>
+#include <Protocol/Smbios.h>
+
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+
+typedef enum {
+  NoButtons     = 0,
+  VolUpButton   = 1,
+  VolDownButton = 2
+} BUTTON_STATE;
+
+#define SMBIOS_VOLUME_UP    "Vol+"
+#define SMBIOS_VOLUME_DOWN  "Vol-"
+
+BUTTON_STATE  gButtonState = NoButtons;
+
+/*
+Say volume up button is pressed because we want to go to frontpage.
+
+@param[in]     - Button Services protocol pointer
+@param[out]    - Pointer to a boolean value to receive the button state
+
+@retval        - EFI_SUCCESS;
+*/
+EFI_STATUS
+EFIAPI
+PreBootVolumeUpButtonThenPowerButtonCheck (
+  IN  MS_BUTTON_SERVICES_PROTOCOL  *This,
+  OUT BOOLEAN                      *PreBootVolumeUpButtonThenPowerButton // TRUE if button combo set else FALSE
+  )
+{
+  DEBUG ((DEBUG_ERROR, "%a \n", __FUNCTION__));
+  *PreBootVolumeUpButtonThenPowerButton = (gButtonState == VolUpButton);
+  return EFI_SUCCESS;
+}
+
+/*
+Say volume down button is pressed because we want alt boot.
+
+@param[in]     - Button Services protocol pointer
+@param[out]    - Pointer to a boolean value to receive the button state
+
+@retval        - EFI_SUCCESS;
+*/
+EFI_STATUS
+EFIAPI
+PreBootVolumeDownButtonThenPowerButtonCheck (
+  IN  MS_BUTTON_SERVICES_PROTOCOL  *This,
+  OUT BOOLEAN                      *PreBootVolumeDownButtonThenPowerButton // TRUE if button combo set else FALSE
+  )
+{
+  DEBUG ((DEBUG_ERROR, "%a \n", __FUNCTION__));
+  *PreBootVolumeDownButtonThenPowerButton = (gButtonState == VolDownButton);
+  return EFI_SUCCESS;
+}
+
+/*
+Clear current button state.
+
+@param[in]     - Button Services protocol pointer
+
+@retval        - EFI_SUCCESS;
+*/
+EFI_STATUS
+EFIAPI
+PreBootClearVolumeButtonState (
+  MS_BUTTON_SERVICES_PROTOCOL  *This
+  )
+{
+  DEBUG ((DEBUG_ERROR, "%a \n", __FUNCTION__));
+  gButtonState = NoButtons;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Get Bios String  - Return address of Bios String
+

```

Comment:
Parameter and return value descriptions needed.

Reply:
Fixed